### DISCUSSION TITLE
Disable auto-select the first item on suggestion

### DISCUSSION THREAD

**Discussion by tranmanhlt7a4** [hrsh7th/nvim-cmp#1411] [Q&A] [answered]
How can I disable auto-select the first item on suggestion until I press a key like <code><TAB>&lt;tab&gt;</code>, <code>&lt;down&gt;</code>, ...?

**Comment by fitrh** [comment, marked as answer]
Set `preselect` to `cmp.PreselectMode.None`, see `:h cmp-config.preselect`

**Reply by tranmanhlt7a4** [reply]
> Set `preselect` to `cmp.PreselectMode.None`, see `:h cmp-config.preselect`

Thank for your help. But that way doesn't work for me until I add 
```lua
complete = {
    completeopt=menu,menuone,noinsert,noselect
},
```
after
```lua 
preselect = require('cmp').PreselectMode.None,
```
**My config file like this:**
```lua
require('cmp').setup({
    -- Default setup in README.md
    preselect = require('cmp').PreselectMode.None,
    complete = {
        completeopt=menu,menuone,noinsert,noselect
    },
})
```

Could you explain why do I have to put that code after <code>preselect</code> to make it work? Thank you very much!

**Reply by hrsh7th** [reply]
The `complete` key is not defined. You should use `completion` instead.

I think you don't need to specify `completion.completeopt`.
You should just specify `set completeopt=menu,menuone,noselect` in general vim way.

**Reply by tranmanhlt7a4** [reply]
Thank you so much!

**Reply by albertfgu** [reply]
I ran into this issue as well. The vim default settings are `completeopt=menu,menuone,noselect`, but the preselect mode was not being set correctly. I also had to specify `completion.completeopt` to fix it. I'm not sure if this is a bug in nvim-cmp

**Reply by Andrew15-5** [reply]
I, too, want to say that adding `completion = { completeopt = "menu,menuone,noselect" }` disables preselection, and it does so without `preselect = cmp.PreselectMode.None`. I'm configuring [NvChad](http://nvchad.com) for the second time and tripping over this problem for the second time as well.

**Reply by Andrew15-5** [reply]
I have an update. I'm using `ray-x/go.nvim` for Go and `mrcjkb/rustaceanvim` for Rust and for some reason after providing the default `capabilities` and `on_attach` they still preselect some entry (although `capabilities` and `on_attach` shouldn't change this behavior). Turns out, adding `preselect = cmp.PreselectMode.None,` fixes the problem. But for all other (manually configured) LSP servers, the `completion = { completeopt = "menu,menuone,noselect" },` is still enough. So the best thing to do is to just write both lines and don't have any problems down the road:

```lua
  preselect = cmp.PreselectMode.None,
  completion = { completeopt = "menu,menuone,noselect" },
```

**Reply by josh-nz** [reply]
@Andrew15-5 refer to the first entry in the nvim-cmp FAQ in the docs, "Why does cmp automatically select a particular item?" which explains why some of your LSPs were still preselecting a completion.

**Reply by rahatarmanahmed** [reply]
What FAQ is this referring to?

I'm having this issue and the config mentioned here doesn't work.

**Reply by rahatarmanahmed** [reply]
Ah, I found it. `:h cmp-faq`

**Comment by Old-Farmer** [comment]
Good Solution!

**Comment by ForsakenRei** [comment]
I don't know why but I cannot get it working. In my `init.vim` the `cmp` section is
```lua
lua <<EOF
  local cmp = require'cmp'

  cmp.setup({
    preselect = cmp.PreselectMode.None,
    completion = {
      completeopt = "menu,menuone,noneselect"
    },
    snippet = {
      expand = function(args)
        vim.fn["vsnip#anonymous"](args.body) -- For `vsnip` users.
      end,
    },
    window = {
      -- completion = cmp.config.window.bordered(),
      -- documentation = cmp.config.window.bordered(),
    },
    mapping = cmp.mapping.preset.insert({
      ['<C-b>'] = cmp.mapping.scroll_docs(-4),
      ['<C-f>'] = cmp.mapping.scroll_docs(4),
      ['<C-Space>'] = cmp.mapping.complete(),
      ['<C-e>'] = cmp.mapping.abort(),
      ['<CR>'] = cmp.mapping.confirm({select = false}),
    }),
    sources = cmp.config.sources({
      { name = 'nvim_lsp' },
      { name = 'vsnip' }, -- For vsnip users.
    }, {
      { name = 'buffer' },
    })
  })

  cmp.setup.filetype('gitcommit', {
    sources = cmp.config.sources({
      { name = 'git' }, 
    }, {
      { name = 'buffer' },
    })
  })

  cmp.setup.cmdline({ '/', '?' }, {
    mapping = cmp.mapping.preset.cmdline(),
    sources = {
      { name = 'buffer' }
    }
  })

  cmp.setup.cmdline(':', {
    mapping = cmp.mapping.preset.cmdline(),
    sources = cmp.config.sources({
      { name = 'path' }
    }, {
      { name = 'cmdline' }
    })
  })
--[[
  -- Set up lspconfig.
  local capabilities = require('cmp_nvim_lsp').default_capabilities()
  -- Replace <YOUR_LSP_SERVER> with each lsp server you've enabled.
  require('lspconfig')['pyright'].setup {
    capabilities = capabilities
  }
  require('lspconfig')['tsserver'].setup {
    capabilities = capabilities
  }
--]]
EOF
```
but I still get the 1st option auto selected?

NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1713484068
Run "nvim -V1 -v" for more info
and latest cmp.

**Reply by Shougo** [reply]
`completeopt = "menu,menuone,noneselect"`

is wrong.

It must be:

`completeopt = "menu,menuone,noselect"`

**Reply by ForsakenRei** [reply]
> `completeopt = "menu,menuone,noneselect"`
> 
> is wrong.
> 
> It must be:
> 
> `completeopt = "menu,menuone,noselect"`

Ahh good catch, thanks! Though after chaging it still felt a little weird. For example, now I type `Plu` and the autocomplete will showup but none of them will be highlighted, which I think the `noselect` is taking effect. However if I hit enter the first option will get completed instead of a new line. Based on search I thought `select = false` will do the trick but actually not...Probably I should just replace `CR` with `Tab` for confirm?

**Reply by Shougo** [reply]
It seems your mapping behavior.

```
      ['<CR>'] = cmp.mapping.confirm({select = false}),
```

**Reply by ForsakenRei** [reply]
> It seems your mapping behavior.
> 
> ```
>       ['<CR>'] = cmp.mapping.confirm({select = false}),
> ```

Yeah I tried to replace `CR` with `Tab` but it still behave the same...? Maybe some other plugin is acting up?

**Reply by Shougo** [reply]
Hm...  

```lua
      ['<CR>'] = cmp.mapping.confirm { select = false },
```

It works for me.  So I think your `<CR>` is mapped to another function.

**Reply by ForsakenRei** [reply]
> Hm...
> 
> ```lua
>       ['<CR>'] = cmp.mapping.confirm { select = false },
> ```
> 
> It works for me. So I think your `<CR>` is mapped to another function.

That's my guess, thanks for all the help, I will inspect my full `init.vim` again carefully.

**Comment by cjnchk** [comment]
Use Supertab config in https://www.lazyvim.org/configuration/recipes, add blow:
<img width="655" alt="image" src="https://github.com/user-attachments/assets/5f194f9d-d311-4eea-aa0c-6d69dd9de7fb">

**Reply by nickjj** [reply]
This didn't work as is, at least not in March 2025.

I uncovered a few things and ultimately ended up opening an issue for this with LazyVim https://github.com/LazyVim/LazyVim/issues/5775.
